### PR TITLE
Natural gradient descent v3

### LINKED ISSUE #894: Natural Gradient Optimizer
Hello, 

I have seen in the [literature](https://arxiv.org/abs/1803.09151) for GP algorithms that use inducing points and variational inference to perform inference can sometimes converge slowly when only using optimizers such as stochastic gradient descent. This is apparently due to the difficulty of optimizing the variational parameters as well as the likelihood and kernel parameters. The authors in that paper suggest that using the **natural gradient optimizer** has shown to have big improvements especially in large-scale GP methods like the **SVGP** and **DeepGP**.

I've seen the NatGrad optimizer has already been implemented in this [branch](https://github.com/cornellius-gp/gpytorch/tree/natural) within the GPyTorch library. I wanted to know what was everyone's experience using it and why is it not in the main branch? Perhaps the nature of using matrix-vector-multiplication methods don't suffer from the same issue of convergence or joint optimization? Or it's just a matter of code coverage?

### PR DESCRIPTION
**Pull request by gpleiss**
Fork of #1207 - final version of @rhaps0dy 's work
[Closes #894]

- [x] Add `NaturalVariationalDistribution`
- [x] Add unit/example tests
- [x] Add tutorial notebook

### CODE REVIEW

**Comment by gpleiss** [comment]
Ready for review @jacobrgardner and/or @martinjankowiak

**Comment by rhaps0dy** [comment]
@gpleiss, thank you so much for properly finishing the natural gradients pull request. I am of the opinion that is very much needed functionality for GPytorch. I'm sorry for not finishing it myself when I started it.

File: gpytorch/variational/natural_variational_distribution.py [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+    parameterized by **natural** parameters.
+
+    .. note::
+       The :obj:`~gpytorch.variational.NaturalVariationalDistribution` can only
+       be used with :obj:`~torch.optim.SGD`, or other optimizers
+       that follow exactly the gradient direction. Failure to do so will cause
+       the natural matrix :math:`\mathbf \Theta_\text{mat}` to stop being
+       positive definite, and a :obj:`~RuntimeError` will be raised.
+
+    .. seealso::
```
Comment (martinjankowiak) [inline gpytorch/variational/natural_variational_distribution.py:30]: typo

File: gpytorch/variational/natural_variational_distribution.py [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+        for use instructions.
+
+        The :obj:`~gpytorch.variational.TrilNaturalVariationalDistribution` for
+        a more numerically stable parameterization, at the cost of needing more
+        iterations to make variational regression converge.
+
+    :param int num_inducing_points: Size of the variational distribution. This implies that the variational mean
+        should be this size, and the variational covariance matrix should have this many rows and columns.
+    :param batch_shape: Specifies an optional batch size
+        for the variational parameters. This is useful for example when doing additive variational inference.
```
Comment (martinjankowiak) [inline gpytorch/variational/natural_variational_distribution.py:42]: what does this mean?

File: examples/04_Variational_and_Approximate_GPs/Natural_Gradient_Descent.ipynb [hunk 3]
```diff
@@ -0,0 +1,376 @@
...
+    "    \n",
+    "    for x_batch, y_batch in minibatch_iter:\n",
+    "        ### Perform NGD step to optimize variational parameters\n",
+    "        variational_ngd_optimizer.zero_grad()\n",
+    "        output = model(x_batch)\n",
+    "        loss = -mll(output, y_batch)\n",
+    "        minibatch_iter.set_postfix(loss=loss.item())\n",
+    "\n",
+    "        # Multiply the loss by the amount of data in order to take the appropriate step size!\n",
+    "        (loss * mll.num_data).backward()\n",
```
Comment (jacobrgardner) [inline examples/04_Variational_and_Approximate_GPs/Natural_Gradient_Descent.ipynb:310]: What do you think about making like a 5 line `class NGD(torch.optim.SGD):` that takes `num_data` and wraps this behavior (e.g., does the multiply on the gradients in step before calling super)? The other primary advantage would be avoiding the trap of using other optimizers, since then we're being pretty explicit about what to use.

File: gpytorch/module.py [hunk 4]
```diff
@@ -159,9 +162,12 @@ def named_constraints(self, memo=None, prefix=""):
         return _extract_named_constraints(module=self, memo=None, prefix="")
 
     def named_variational_parameters(self):
-        for name, param in self.named_parameters():
-            if "variational_" in name:
```
Comment (jacobrgardner) [inline gpytorch/module.py:163]: hooray 🙌

File: examples/04_Variational_and_Approximate_GPs/Natural_Gradient_Descent.ipynb (hunk 3 above)
Comment (gpleiss) [reply examples/04_Variational_and_Approximate_GPs/Natural_Gradient_Descent.ipynb:310]: Yeah. That's actually a really good idea.

File: gpytorch/module.py (hunk 4 above)
Comment (gpleiss) [reply gpytorch/module.py:163]: Please do not insult the best 2 lines of code I've ever written.

**Comment by gpleiss** [comment]
No worries @rhaps0dy - thanks for all your work on it!

File: gpytorch/variational/natural_variational_distribution.py (hunk 1 above)
Comment (gpleiss) [reply gpytorch/variational/natural_variational_distribution.py:30]: Sorry, where?

File: gpytorch/variational/natural_variational_distribution.py (hunk 2 above)
Comment (gpleiss) [reply gpytorch/variational/natural_variational_distribution.py:42]: I'll remove that comment.

File: examples/04_Variational_and_Approximate_GPs/Natural_Gradient_Descent.ipynb (hunk 3 above)
Comment (gpleiss) [reply examples/04_Variational_and_Approximate_GPs/Natural_Gradient_Descent.ipynb]: Okay - pushed.

**Comment by rhaps0dy** [comment]
Note: it is possible to use other optimizers for the `TrilNaturalVariationalDistribution`. That one's parameterization is a lower triangular matrix for the covariance.

**Comment by gpleiss** [comment]
Question for @rhaps0dy / @jacobrgardner / @martinjankowiak - for the NGD optimization loop we have two options:

1) take alternating steps for NGD/variational parameters + Adam/hyperparameters. (I believe [this is what GPFlow does](https://gpflow.readthedocs.io/en/master/notebooks/advanced/natural_gradients.html#Comparison-with-ordinary-gradients-in-the-conjugate-case)). e.g:
```python
    for x_batch, y_batch in minibatch_iter:
        ### Perform NGD step to optimize variational parameters
        variational_ngd_optimizer.zero_grad()
        output = model(x_batch)
        loss = -mll(output, y_batch)
        minibatch_iter.set_postfix(loss=loss.item())
        loss.backward()
        variational_ngd_optimizer.step()

        ### Perform Adam step to optimize hyperparameters
        hyperparameter_optimizer.zero_grad()
        output = model(x_batch)
        loss = -mll(output, y_batch)
        loss.backward()
        hyperparameter_optimizer.step()
```

2) take the NGD step and the Adam step simultaneously - e.g:
```python
    for x_batch, y_batch in minibatch_iter:
        variational_ngd_optimizer.zero_grad()
        hyperparameter_optimizer.zero_grad()
        output = model(x_batch)
        loss = -mll(output, y_batch)
        minibatch_iter.set_postfix(loss=loss.item())
        loss.backward()
        variational_ngd_optimizer.step()
        hyperparameter_optimizer.step()
```

Thoughts?

**Comment by gpleiss** [comment]
@rhaps0dy - probably best to encourage people to still use the NGD optimizer for `TrilNaturalVariationalStrategy`, right? I imagine that trying to combine adaptive gradients with NGD could cause problems.

@jacobrgardner @martinjankowiak - went with joint NGD/Adam step rather than alternating. At this point I think the PR is ready for a final review.

**Comment by jacobrgardner** [comment]
@gpleiss the paper specifies that they alternate: https://arxiv.org/abs/1803.09151 (section 5.3).

Might be good to have this be the default for drop in reproducibility? We could do a study to see if joint was better in practice

**Comment by gpleiss** [comment]
@jacobrgardner - on elevators it seems to produce the same result. @martinjankowiak suggested joint because its more efficient.

**Comment by gpleiss** [comment]
The example notebook shows both options though.
